ENH: omit CSSStyle rules for hidden table elements (#43619)

rendner opened the pull request:
- [x] closes #43619
- [x] tests added / passed (but canceled due to another error)
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

I couldn't run all tests, because of the following problem, which isn't related to my changes:
```
pandas/tests/io/test_orc.py terminate called after throwing an instance of 'orc::TimezoneError'
  what():  Can't open /etc/localtime
Fatal Python error: Aborted
```

rendner committed:
ENH: omit CSSStyle rules for hidden table cells (#43619)

attack68 commented on doc/source/whatsnew/v1.3.4.rst:
```diff
@@ -33,10 +33,9 @@ Bug fixes
-~~~~~
--
--
+Performance improvements
```
this won't make it for 1.3.4, will have to place in 1.4.0. Can add the note to the `Styler` section in that one.

rendner replied:
done

attack68 commented on pandas/tests/io/formats/style/test_html.py:
```diff
@@ -467,3 +467,42 @@ def test_maximums(styler_mi, rows, cols):
+
+def test_omit_css_style_rules_for_hidden_cells():
+    # GH 43619
+    df = DataFrame([[1, 2], [3, 4]], columns=["A", "B"])
```
can you maybe use the fixtures that are available in the module `styler` or `styler_mi`?

rendner replied:
done

attack68 commented on pandas/tests/io/formats/style/test_html.py:
```diff
@@ -467,3 +467,42 @@ def test_maximums(styler_mi, rows, cols):
+          color: #f1f1f1;
+        }
+        </style>
+        <table id="T_">
```
everything below `</style>` is already tested. To minimise and make the test more specific just try `assert expected in result` where `expected = <style>...</style>`

rendner replied:
done

attack68 reviewed (changes requested):
looks good but comments

attack68 commented on pandas/tests/io/formats/style/test_html.py:
```diff
@@ -467,3 +467,42 @@ def test_maximums(styler_mi, rows, cols):
+    df = DataFrame([[1, 2], [3, 4]], columns=["A", "B"])
+    result = (
+        df.style.set_uuid("")
+        .background_gradient()
```
also don't use `background_gradient` since it requires matplotlib which is not an available library in this test modules in some of the test environments.

use `applymap(lambda v: "color: blue;")`: its also easier to follow.

rendner replied:
done

attack68 commented:
You don't have to, but might also like to try and solve the same issue for the index or column headers, that might get CSS for ids if `applymap_index(lambda v: "color: blue;")` is used..

rendner committed:
simplify test

- use fixture
- test only styles
- use "applymap" instead of "background_gradient"

rendner committed:
omit CSSStyle rules for hidden index labels

rendner committed:
omit CSSStyle rules for hidden column labels

rendner committed:
move "whatsnew entry" to "1.4.0" release notes

rendner commented:
> You don't have to, but might also like to try and solve the same issue for the index or column headers, that might get CSS for ids if `applymap_index(lambda v: "color: blue;")` is used..

I also implemented it for column and index headers.

jreback commented:
do we have an asv for this? can you show the perf change (and add asv if we don't have).

attack68 commented:
> do we have an asv for this? can you show the perf change (and add asv if we don't have).
I will have a look at this and make a suggestion for a more general asv that can also test this.
Note that this can also improve browser performance as well, not just pandas performance.

attack68 reviewed (approved):
lgtm. i will take a look at the asvs, but even if the performance improvement is marginal it makes better HTML anyway.

jreback commented:
> lgtm. i will take a look at the asvs, but even if the performance improvement is marginal it makes better HTML anyway.

sure, just covering bases

attack68 commented:
@rendner you can use the code in #43687 to see the rendering performance enhancement. I expect minimal, but the performance in HTML might be better for large dataframes

rendner commented:
@attack68 Thank you.

>asv continuous -f 1.1 upstream/master HEAD -b style.Render -E virtualenv

The command gave me the following error `Unknown commit upstream/master`.
So I used this one `asv continuous -f 1.1 011c016f6 HEAD -b style.Render -E virtualenv`.
`011c016f6` is the first commit before my changes, which you can verify here: https://github.com/rendner/pandas/commits/enh-43619

Here are the results:
```
       before           after         ratio
     [011c016f]       [fe9e44b9]
     <master>         <enh-43619>
-      21.1±0.2ms       19.0±0.3ms     0.90  io.style.Render.time_apply_format_hide_render(36, 12)
-      75.5±0.8ms       61.6±0.6ms     0.82  io.style.Render.time_apply_format_hide_render(24, 120)
-      40.7±0.4ms       33.0±0.7ms     0.81  io.style.Render.time_apply_format_hide_render(12, 120)
-       110±0.5ms       88.1±0.4ms     0.80  io.style.Render.time_apply_format_hide_render(36, 120)

SOME BENCHMARKS HAVE CHANGED SIGNIFICANTLY.
PERFORMANCE INCREASED.
```
I guess the performance for hidden cells could be further improved by moving the [data_element_visible](https://github.com/rendner/pandas/blob/enh-43619/pandas/io/formats/style_render.py#L596) check at the start of the for loop. And if not visible continue with the next cell. But I'm unsure how much we would gain.

attack68 commented:
> 
> I guess the performance for hidden cells could be further improved by moving the [data_element_visible](https://github.com/rendner/pandas/blob/enh-43619/pandas/io/formats/style_render.py#L596) check at the start of the for loop. And if not visible continue with the next cell. But I'm unsure how much we would gain.

This is good, don't need anything else. Im surprised it gets this much improvement but the asv does specifically hide almost all elements. @jreback happy to merge?

jreback reviewed (approved):
lgtm

attack68 commented:
thanks @rendner
